proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8972,134 +8987,101 @@ lemma refill_budget_check_valid_ready_qs_not_queued:
-
-  \<comment> \<open>the useful fact, again\<close>
-  apply (prop_tac "unat (cur_time s + kernelWCET_ticks)
-                   = unat (cur_time s) + unat kernelWCET_ticks")
-   apply (rule unat_add_lem')
-   using cur_time_no_overflow no_olen_add_nat apply blast
-
-  apply (intro conjI; intro impI)
-
-   \<comment> \<open>tl (sc_refills sc) = []\<close>
-
-   apply (prop_tac "length (sc_refills sc) = Suc 0")
-    apply (clarsimp simp: valid_refills_def rr_valid_refills_def vs_all_heap_simps MIN_REFILLS_def)
-    apply (metis Nitpick.size_list_simp(2))
-
-   apply (prop_tac "sc_budget sc = r_amount (refill_hd sc)")
-    apply (clarsimp simp: valid_refills_def rr_valid_refills_def vs_all_heap_simps)
-    apply (case_tac "sc_refills sc", force)
-    apply (clarsimp split: if_splits)
-
-   apply (intro conjI impI
-          ; (clarsimp simp: valid_refills_def rr_valid_refills_def vs_all_heap_simps MIN_REFILLS_def MIN_SC_BUDGET_def)?)
-    apply ((simp add: word_le_nat_alt, blast intro: order_trans unat_plus_gt)+)[2]
-
-  \<comment> \<open>tl (sc_refills sc) \<noteq> []\<close>
-
-  apply (simp add: word_le_nat_alt, blast intro: order_trans unat_plus_gt)
+   apply (clarsimp simp: valid_refills_def rr_valid_refills_def vs_all_heap_simps obj_at_def)
+   apply (clarsimp simp: word_le_nat_alt word_less_nat_alt unat_sub)
+   apply (subst refills_merge_prefix_r_time_hd)
+   apply (subst MIN_BUDGET_merge_same_hd_r_time)
+     apply (fastforce simp: schedule_used_non_nil)
+    apply (subst xtr6[OF schedule_used_r_amount_head]; fastforce?)
+     apply clarsimp
+     apply (frule (2) unat_sum_list_equals_budget)
+      apply (clarsimp simp: word_le_nat_alt)
+     apply (metis (mono_tags, hide_lams) refill_budget_check_word_helper Groups.add_ac(1)
+                  unat_bounded_above)
+    apply clarsimp
+    apply (prop_tac "min usage (r_amount (refill_hd y)) = usage")
+     using unat_mono word_le_not_less apply fastforce
+    apply (metis linorder_min_same1 unat_sub word_le_nat_alt)
+   apply (subst schedule_used_hd_r_time_same; fastforce?)
+    apply (metis MIN_REFILLS_def Nitpick.size_list_simp(2) One_nat_def Suc_eq_plus1 Suc_le_eq
+                 list.size(4) numeral_eqs(1))
+   apply clarsimp
+   apply (prop_tac "min usage (r_amount (refill_hd y)) = usage")
+    using unat_mono word_le_not_less apply fastforce
+   apply (metis (mono_tags) Groups.add_ac(2) One_nat_def add_diff_cancel_right' cur_time_no_overflow
+                current_time_bounded_strengthen le0 unat_plus_gt_trans unat_plus_simple word_sub_le)
+  apply (clarsimp simp: sc_valid_refills_def split: if_splits)
+   apply (clarsimp simp: obj_at_def)
+  apply (clarsimp simp: obj_at_def)
+  apply (case_tac "0 < usage")
+   using MIN_BUDGET_pos apply simp
+  apply (clarsimp simp: refill_ready_def refill_ready_no_overflow_def)
+  apply (clarsimp simp: current_time_bounded_def word_le_nat_alt)
+  apply (metis add_leE unat_plus_simple unat_sum_bound_equiv)
   done
 
```

Comment:
As discussed in the MCS meeting today, we divide the reviewing task of this file in two, suggesting here for the division point. For the sake of simplifying the process, I would like to request reviews from:

Corey, Victor
----------< division point
Ed, Matt

For others, please add names to either side if you are joining the review. I’ll look though the file anyway. Thanks.